avogadro/qtplugins/surfaces/surfaces.cpp

```diff
@@ -183,69 +186,149 @@ void Surfaces::calculateSurface()
-
-    double minDistance = mdist;
-    if (neighbors == nullptr)
-      neighbors = new Array<Index>;
-    neighborPerceiver.getNeighborsInclusiveInPlace(*neighbors, pos);
-    for (Index neighbor: *neighbors) {
-      Vector3 neighborPos = m_molecule->atomPosition3d(neighbor);
-      double distance = (neighborPos - pos).norm();
-      distance -= radii[neighbor] + radiusOffset;
-      minDistance = std::min(minDistance, distance);
+    // first, make a list of all atom positions and radii
+    Array<Vector3> atomPositions = m_molecule->atomPositions3d();
+    std::vector<std::pair<Vector3, double>> *atoms =
+      new std::vector<std::pair<Vector3, double>>(m_molecule->atomCount());
+    double max_radius = probeRadius;
+    for (size_t i = 0; i < atoms->size(); i++) {
+      (*atoms)[i].first = atomPositions[i];
+      (*atoms)[i].second = Core::Elements::radiusVDW(m_molecule->atomicNumber(i)) + probeRadius;
+      if ((*atoms)[i].second > max_radius)
+        max_radius = (*atoms)[i].second;
     }
 
-    m_cube->setValue(in(0), in(1), in(2), minDistance);
+    double padding = max_radius + probeRadius;
+    m_cube->setLimits(*m_molecule, m_dialog->resolution(), padding);
+    Vector3i size = m_cube->dimensions();
+    for (int z = 0; z < size(2); z++)
+      for (int y = 0; y < size(1); y++)
+        for (int x = 0; x < size(0); x++)
+          m_cube->setValue(x, y, z, -1.0f);
+
+    const float res = m_dialog->resolution();
+    const Vector3 min = m_cube->min();
+    const float mdist = probeRadius;
+
+    // then, for each atom, set cubes around it up to a certain radius
+    QFuture innerFuture = QtConcurrent::map(*atoms, [=](std::pair<Vector3, double> &in) {
+      double startPosZ = in.first(2) - in.second;
+      double endPosZ = in.first(2) + in.second;
+      int startIndexZ = (startPosZ - min(2)) / res;
+      int endIndexZ = (endPosZ - min(2)) / res + 1;
+      for (int indexZ = startIndexZ; indexZ < endIndexZ; indexZ++) {
+        double posZ = indexZ * res + min(2);
+        double radiusZsq = pow(in.second, 2) - pow(posZ - in.first(2), 2);
+        if (radiusZsq < 0.0)
+          continue;
+        double radiusZ = sqrt(radiusZsq);
+        double startPosY = in.first(1) - radiusZ;
+        double endPosY = in.first(1) + radiusZ;
+        int startIndexY = (startPosY - min(1)) / res;
+        int endIndexY = (endPosY - min(1)) / res + 1;
+        for (int indexY = startIndexY; indexY < endIndexY; indexY++) {
+          double posY = indexY * res + min(1);
+          double lengthZYsq = pow(radiusZ, 2) - pow(posY - in.first(1), 2);
+          if (lengthZYsq < 0.0)
+            continue;
+          double lengthZY = sqrt(lengthZYsq);
+          double startPosX = in.first(0) - lengthZY;
+          double endPosX = in.first(0) + lengthZY;
+          int startIndexX = (startPosX - min(0)) / res;
+          int endIndexX = (endPosX - min(0)) / res + 1;
+          for (int indexX = startIndexX; indexX < endIndexX; indexX++) {
```

Comment:
I don't remember whether the `Cube` class stores strides in x, y, z.. (And it *definitely* needs documentation / comments on that.)

But this could also be a case for a `Cube::fill()` if, like I assume, changes in `x` values are contiguous.

If instead, it's ordered differently, ignore this comment.
(I don't have time tonight to skim the code..)

Reply:
It looks like `Cube` stores values contiguous in `z`, so I could reorder those loops and use an overloaded `Cube::fill()` as you mention.